### PR TITLE
[fix #14] Use IOTA as a cache to cache and pre-verify transactions

### LINKED ISSUE #14: Use IOTA as a cache to cache and pre-verify transactions

### PR DESCRIPTION
**Pull request by stplaydog**

### CODE REVIEW

File: scripts/iota_cache.py [hunk 1]
```diff
@@ -0,0 +1,67 @@
+import time
+import sys
+from iota import Iota, Address, ProposedTransaction, Tag, Transaction, TryteString, TransactionTrytes, ProposedBundle
+from six import binary_type, moves as compat, text_type
+
+
+class IotaCache:
+    def __init__(self, uri=None, seed=None):
+        if not uri:
+            self.uri = "http://localhost:14700"
```
Comment (wunder3605) [inline scripts/iota_cache.py:10]:
```
else:
    self.uri = uri
```

File: scripts/iota_cache.py [hunk 2]
```diff
@@ -0,0 +1,67 @@
...
+        if not seed:
+            self.seed = 'EBZYNR9YVFIOAZUPQOLRZXPPPIKRCJ9EJKVCXMYVLMNOCCOPYPJKCWUZNLJZZZZWTMVQUXZFYLVLZXJ9Q'
```
Comment (wunder3605) [inline scripts/iota_cache.py:14]:
```
else:
    self.seed = seed
```

File: scripts/iota_cache.py [hunk 3]
```diff
@@ -0,0 +1,67 @@
...
+        self.api = Iota(self.uri, self.seed)
+
+    def cacheTxnInTangle(self, IPFSAddr, tag):
```
Comment (wunder3605) [inline scripts/iota_cache.py]: May be use cache_txn_in_tangle to name the this method is better?
Also use ipfs_addr to name the parameter is better?

File: scripts/iota_cache.py [hunk 4]
```diff
@@ -0,0 +1,67 @@
...
+        for txnTrytes in txnTrytesAll['trytes']:
+            txn = Transaction.from_tryte_string(txnTrytes)
+            if txn.signature_message_fragment not in consumedSet:
+                msgTryte = txn.signature_message_fragment.decode()
+                ret.append(msgTryte)
+        return ret
+
+    def setTxnAsSynced(self, IPFSAddr, tag):
+        result = self.cacheTxnInTangle(IPFSAddr, tag+b"CONSUMED")
+        return result
```
Comment (wunder3605) [inline scripts/iota_cache.py:69]: Add TagHelper class
```
class TagHelper(object):

    TAG_YEAR = {
        2016: b"B9AF",
        2017: b"B9AG",
        2018: b"B9AH",
        2019: b"B9AI",
        2020: b"B9AH",
    }
    TAG_MONTH = [
        b"99",                              # 00
        b"9A", b"9B", b"9C", b"9D", b"9E",  # 01-05
        b"9F", b"9G", b"9H", b"9I", b"A9",  # 06-10
        b"AA", b"AB"                        # 11-12
    ]
    TAG_DAY = [
        b"99",                              # 00
        b"9A", b"9B", b"9C", b"9D", b"9E",  # 01-05
        b"9F", b"9G", b"9H", b"9I", b"A9",  # 06-10
        b"AA", b"AB", b"AC", b"AD", b"AE",  # 11-15
        b"AF", b"AG", b"AH", b"AI", b"B9",  # 16-20
        b"BA", b"BB", b"BC", b"BD", b"BE",  # 21-25
        b"BF", b"BG", b"BH", b"BI", b"C9",  # 26-30
        b"CA",                              # 31
    ]

    @classmethod
    def get_current_tag(cls):
        today = datetime.date.today()
        return TagHelper.TAG_YEAR[today.year] + \
            TagHelper.TAG_MONTH[today.month] + \
            TagHelper.TAG_DAY[today.day]

    @classmethod
    def get_previous_tag(cls):
        today = datetime.date.today()
        one_day = datetime.timedelta(days=1)
        yesterday = today - one_day
        return TagHelper.TAG_YEAR[yesterday.year] + \
            TagHelper.TAG_MONTH[yesterday.month] + \
            TagHelper.TAG_DAY[yesterday.day]
```

File: scripts/iota_cache.py [hunk 5]
```diff
@@ -0,0 +1,67 @@
+import time
+import sys
+from iota import Iota, Address, ProposedTransaction, Tag, Transaction, TryteString, TransactionTrytes, ProposedBundle
+from six import binary_type, moves as compat, text_type
+
+
+class IotaCache:
```
Comment (wunder3605) [inline scripts/iota_cache.py]:
```
class IotaCache(object):
```
Inherit from object is recommended in Python.

File: scripts/iota_cache.py [hunk 6]
```diff
@@ -0,0 +1,67 @@
+import time
+import sys
```
Comment (wunder3605) [inline scripts/iota_cache.py]: Remove unused import.

File: scripts/iota_cache.py [hunk 7]
```diff
@@ -0,0 +1,67 @@
+import time
```
Comment (wunder3605) [inline scripts/iota_cache.py:1]: Remove unused import.

File: scripts/iota_cache_test.py [hunk 8]
```diff
@@ -0,0 +1,74 @@
+#from unittest.mock import MagicMock
+import mock
+from mock import MagicMock
+import unittest
+from iota import TransactionHash, Iota, Address, ProposedTransaction, Tag, Transaction, TryteString, TransactionTrytes, ProposedBundle
```
Comment (wunder3605) [inline scripts/iota_cache_test.py]: Need split this line to two lines.

File: scripts/iota_cache.py (hunk 7 above)
Comment (wunder3605) [inline scripts/iota_cache.py:1]: Add "import datetime" for TagHelper class.

File: scripts/iota_cache.py [hunk 9]
```diff
@@ -0,0 +1,67 @@
...
+        tips = self.api.get_tips()
+        states = self.api.get_inclusion_states(transactions['hashes'], tips['hashes'])
+        i = 0
+        for txn in transactions['hashes']:
+            if states['states'][i] == True:
+                ret.append(txn)
+        return ret
+
+    def getNonConsumedTxns(self, tag):
+        txnHashesAll = self.getApprovedTxns(tag)
```
Comment (wunder3605) [inline scripts/iota_cache.py]: Variable in method should be lowercase.

File: scripts/iota_cache.py [hunk 10]
```diff
@@ -0,0 +1,67 @@
...
+        print txnHashesAll
```
Comment (wunder3605) [inline scripts/iota_cache.py]: Use "logging.debug" instead of "print" is better?

File: scripts/iota_cache.py [hunk 11]
```diff
@@ -0,0 +1,67 @@
...
+            txn = Transaction.from_tryte_string(txnTrytes)
+            consumedSet.append(txn.signature_message_fragment)
+        for txnTrytes in txnTrytesAll['trytes']:
+            txn = Transaction.from_tryte_string(txnTrytes)
+            if txn.signature_message_fragment not in consumedSet:
+                msgTryte = txn.signature_message_fragment.decode()
+                ret.append(msgTryte)
+        return ret
+
+    def setTxnAsSynced(self, IPFSAddr, tag):
```
Comment (wunder3605) [inline scripts/iota_cache.py]: IPFSAddr -> ipfs_addr?

File: scripts/iota_cache.py [hunk 12]
```diff
@@ -0,0 +1,67 @@
...
+    def getApprovedTxns(self, tag):
+        ret = []
+        transactions = self.api.find_transactions(None, None, [tag], None)
+        if len(transactions['hashes']) == 0:
+            return ret
+        tips = self.api.get_tips()
+        states = self.api.get_inclusion_states(transactions['hashes'], tips['hashes'])
+        i = 0
+        for txn in transactions['hashes']:
+            if states['states'][i] == True:
```
Comment (wunder3605) [inline scripts/iota_cache.py]: if states['states'][i] == True:
May be in this way is recommended:
if states['states'][i] is True:
or
if states['states'][i]:

File: scripts/iota_cache_test.py [hunk 13]
```diff
@@ -0,0 +1,74 @@
...
+
+
+from iota_cache import IotaCache
+
+# test Count class
```
Comment (wunder3605) [inline scripts/iota_cache_test.py]: Recommend 2 blank lines before class in PEP 8.

File: scripts/iota_cache_test.py [hunk 14]
```diff
@@ -0,0 +1,74 @@
...
+
+    def mock_getApprovedTxns(self, tag):
+        ret = []
+        for txn in self.transfers:
+            if txn[0].tag == Tag(tag):
+                ret.append(txn[0].hash)
+        return ret
+
+    def mock_find_transactions(self, bundles, address, tags, approvees):
+        ret = {'hashes' : [] }
```
Comment (wunder3605) [inline scripts/iota_cache_test.py]: {'hashes': []}

File: scripts/iota_cache_test.py [hunk 15]
```diff
@@ -0,0 +1,74 @@
...
+        for txn in self.transfers:
+            if txn[0].tag == Tag(tags[0]):
+                ret['hashes'].append(txn[0].hash)
+        return ret
+
+    def mock_get_trytes(self, txnHashesAll):
+        ret = {'trytes' :[]}
```
Comment (wunder3605) [inline scripts/iota_cache_test.py]: {'trytes': []}

File: scripts/iota_cache_test.py [hunk 16]
```diff
@@ -0,0 +1,74 @@
...
+                if txn[0].hash == txnHash:
+                    bundle = ProposedBundle()
+                    bundle.add_transaction(txn[0])
+                    bundle.finalize()
+                    ret['trytes'].append(bundle.as_tryte_strings()[0])
+        return ret
+
+    def test_cacheTxnInTangle(self):
+        cache = IotaCache()
+        cache.api.get_new_addresses = MagicMock(return_value={'addresses' : [b'KA9XBBDIYSTVOTEMAUL9JBYKWDN9WAJQWJO9GQYBNOJBSVRBESQXYETFDCSFFWL9PPSPWWAGSNMQSSMBC']})
```
Comment (wunder3605) [inline scripts/iota_cache_test.py]:
```
        cache.api.get_new_addresses = MagicMock(
            return_value={
                'addresses': [b'KA9XBBDIYSTVOTEMAUL9JBYKWDN9WAJQWJO9GQYBNOJBSVRBESQXYETFDCSFFWL9PPSPWWAGSNMQSSMBC']
            }
        )
```

File: scripts/iota_cache_test.py [hunk 17]
```diff
@@ -0,0 +1,74 @@
...
+        cache.api.send_transfer = MagicMock(return_value='OK')
+        ret = cache.cacheTxnInTangle('addr1', 'TAG')
+        self.assertEqual(ret, 'OK')
+
+    def test_getNonConsumedTxns(self):
+        cache = IotaCache()
+        cache.api.get_new_addresses = MagicMock(return_value={'addresses' : [b'KA9XBBDIYSTVOTEMAUL9JBYKWDN9WAJQWJO9GQYBNOJBSVRBESQXYETFDCSFFWL9PPSPWWAGSNMQSSMBC']})
```
Comment (wunder3605) [inline scripts/iota_cache_test.py]:
```
        cache.api.get_new_addresses = MagicMock(
            return_value={
                'addresses': [b'KA9XBBDIYSTVOTEMAUL9JBYKWDN9WAJQWJO9GQYBNOJBSVRBESQXYETFDCSFFWL9PPSPWWAGSNMQSSMBC']
            }
        )
```

File: scripts/iota_cache_test.py [hunk 18]
```diff
@@ -0,0 +1,74 @@
...
+        ret = cache.cacheTxnInTangle('addr8', 'TAG')
+        ret = cache.cacheTxnInTangle('addr3', 'TAGCONSUMED')
+        ret = cache.cacheTxnInTangle('addr4', 'TAGCONSUMED')
+        ret = cache.cacheTxnInTangle('addr5', 'TAGCONSUMED')
+        cache.getApprovedTxns = MagicMock(side_effect=self.mock_getApprovedTxns)
+        cache.api.get_trytes = MagicMock(side_effect=self.mock_get_trytes)
+        cache.api.find_transactions = MagicMock(side_effect=self.mock_find_transactions)
+        print cache.getNonConsumedTxns('TAG')
+
+if __name__ == '__main__':
```
Comment (wunder3605) [inline scripts/iota_cache_test.py:80]: Recommend 2 blank lines after class in PEP 8.

File: scripts/iota_cache_test.py [hunk 19]
```diff
@@ -0,0 +1,74 @@
...
+    unittest.main()
```
Comment (wunder3605) [inline scripts/iota_cache_test.py:81]: In PEP 8, recommend to add newlines at the end of file.

**Review by wunder3605 — CHANGES_REQUESTED** [review]

File: scripts/iota_cache.py (hunk 6 above)
Comment (stplaydog) [reply scripts/iota_cache.py]: done

File: scripts/iota_cache.py (hunk 4 above)
Comment (stplaydog) [reply scripts/iota_cache.py:69]: I think this class should be added separately

File: scripts/iota_cache.py (hunk 7 above)
Comment (stplaydog) [reply scripts/iota_cache.py:1]: This logic should be separated from this class

**Review by wunder3605 — APPROVED** [review]
+1

**Pull request merged**